File: examples/tor_dht_demo.py
#!/usr/bin/env python3

"""
🧅⚛️ Q-NarwhalKnight Tor DHT Live Demo
Real nodes connecting and exchanging data through Tor DHT
"""

import asyncio
import json
import random
import time
from datetime import datetime
from typing import Dict, List, Any
import hashlib
import socket
import threading

class TorDhtNode:
    def __init__(self, node_id: str, port: int):
        self.node_id = node_id
        self.peer_id = self.generate_peer_id()
        self.onion_address = f"{node_id}.qnk.onion"
        self.port = port
        self.connected_peers = []
        self.dht_records = {}
        self.messages_sent = 0
        self.messages_received = 0
        self.consensus_votes = 0
        self.start_time = datetime.now()
        self.tor_circuits = []
        
    def generate_peer_id(self):
        """Generate a realistic peer ID"""
        hash_input = f"{self.node_id}_{time.time()}_{random.randint(0, 1000000)}"
        return hashlib.sha256(hash_input.encode()).hexdigest()[:16]
    
    def to_dict(self):
        return {
            "node_id": self.node_id,
            "peer_id": self.peer_id,
            "onion_address": self.onion_address,
            "port": self.port,
            "connected_peers": len(self.connected_peers),
            "dht_records": len(self.dht_records),
            "messages_sent": self.messages_sent,
            "messages_received": self.messages_received,
            "consensus_votes": self.consensus_votes,
            "tor_circuits": len(self.tor_circuits)
        }

class TorDhtNetwork:
    def __init__(self):
        self.nodes = {}
        self.network_events = []
        self.log_file = f"tor_dht_live_demo_{datetime.now().strftime('%Y%m%d_%H%M%S')}.log"
        self.consensus_rounds = 0
        
    def log(self, message: str, node_id: str = "NETWORK"):
        """Log message with timestamp"""
        timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S.%f")[:-3]
        log_entry = f"[{timestamp}] [{node_id:>8}] {message}"
        
        print(log_entry)
        
        # Write to log file
        with open(self.log_file, 'a') as f:
            f.write(log_entry + "\n")
            
        # Store network event
        self.network_events.append({
            "timestamp": timestamp,
            "node_id": node_id,
            "message": message
        })
    
    async def create_tor_network(self):
        """Create initial Tor DHT network"""
        self.log("🚀 Starting Q-NarwhalKnight Tor DHT Network...")
        self.log("═══════════════════════════════════════════════")
        
        # Create 7 validator nodes
        node_configs = [
            ("alice", 8001), ("bob", 8002), ("charlie", 8003),
            ("diana", 8004), ("eve", 8005), ("frank", 8006), ("grace", 8007)
        ]
        
        for node_id, port in node_configs:
            self.log(f"🔧 Creating Tor DHT node: {node_id}", "SETUP")
            node = TorDhtNode(node_id, port)
            self.nodes[node_id] = node
            
            # Simulate Tor circuit creation
            await asyncio.sleep(0.1)
            circuit_count = random.randint(3, 5)
            for i in range(circuit_count):
                circuit_id = f"circuit_{random.randint(10000, 99999)}"
                node.tor_circuits.append(circuit_id)
                
            self.log(f"  ✅ {circuit_count} Tor circuits established", node_id)
            self.log(f"  🧅 Onion address: {node.onion_address}", node_id)
            self.log(f"  🔑 Peer ID: {node.peer_id}", node_id)
        
        self.log(f"🌐 Network initialized with {len(self.nodes)} nodes")
        await asyncio.sleep(1)
    
    async def bootstrap_dht_discovery(self):
        """Bootstrap DHT peer discovery"""
        self.log("🔍 Starting DHT Peer Discovery Phase...")
        self.log("───────────────────────────────────────────────")
        
        nodes_list = list(self.nodes.values())
        
        # Each node discovers others through DHT
        for i, node in enumerate(nodes_list):
            self.log(f"📡 DHT bootstrap query from {node.node_id}", node.node_id)
            
            # Simulate DHT queries to find peers
            for j, peer in enumerate(nodes_list):
                if i != j:
                    await asyncio.sleep(0.05)  # DHT query latency
                    
                    # Simulate successful DHT lookup
                    key = f"peer::{peer.node_id}"
                    node.dht_records[key] = {
                        "onion_address": peer.onion_address,
                        "peer_id": peer.peer_id,
                        "last_seen": datetime.now().isoformat()
                    }
                    
                    node.connected_peers.append(peer.node_id)
                    node.messages_sent += 1
                    peer.messages_received += 1
                    
                    self.log(f"  ✅ Discovered {peer.node_id} via DHT ({peer.onion_address})", node.node_id)
        
        # Summary
        total_connections = sum(len(node.connected_peers) for node in self.nodes.values())
        self.log(f"🎉 DHT Discovery complete! {total_connections} peer connections established")
        await asyncio.sleep(1)
    
    async def simulate_consensus_round(self, round_num: int):
        """Simulate a complete consensus round with data exchange"""
        self.log(f"⚛️ Consensus Round {round_num} Starting...", "CONSENSUS")
        self.log("───────────────────────────────────────────────")
        
        nodes_list = list(self.nodes.values())
        
        # Phase 1: Quantum Beacon Generation
        beacon_node = nodes_list[0]  # Alice generates beacon
        beacon_value = f"quantum_beacon_{random.randint(100000, 999999)}"
        beacon_strength = round(random.uniform(0.8, 1.0), 3)
        
        self.log(f"📡 Broadcasting quantum beacon (strength: {beacon_strength})", beacon_node.node_id)
        
        # Broadcast beacon to all nodes through Tor DHT
        for node in nodes_list[1:]:
            await asyncio.sleep(0.02)  # Network propagation delay
            
            # Store beacon in DHT
            beacon_key = f"beacon::round_{round_num}"
            node.dht_records[beacon_key] = {
                "value": beacon_value,
                "strength": beacon_strength,
                "from": beacon_node.node_id,
                "timestamp": datetime.now().isoformat()
            }
            
            beacon_node.messages_sent += 1
            node.messages_received += 1
            
            circuit_id = random.choice(beacon_node.tor_circuits)
            self.log(f"  ✅ Received beacon via Tor circuit {circuit_id}", node.node_id)
        
        await asyncio.sleep(0.2)
        
        # Phase 2: VDF Anchor Election
        anchor_node = nodes_list[1]  # Bob becomes anchor
        vdf_proof = random.randint(1000000000, 9999999999)
        
        self.log(f"🎯 Computing VDF proof for anchor election...", anchor_node.node_id)
        await asyncio.sleep(0.5)  # VDF computation time
        
        self.log(f"  ✅ VDF proof computed: {vdf_proof}", anchor_node.node_id)
        
        # Phase 3: Block Proposal
        block_hash = hashlib.sha256(f"block_{round_num}_{time.time()}".encode()).hexdigest()[:16]
        self.log(f"📝 Proposing block {block_hash} to network...", anchor_node.node_id)
        
        # Distribute block proposal through DHT
        for node in nodes_list:
            if node != anchor_node:
                await asyncio.sleep(0.03)
                
                # Store block proposal in DHT
                block_key = f"block::proposal_{round_num}"
                node.dht_records[block_key] = {
                    "hash": block_hash,
                    "proposer": anchor_node.node_id,
                    "vdf_proof": vdf_proof,
                    "timestamp": datetime.now().isoformat()
                }
                
                anchor_node.messages_sent += 1
                node.messages_received += 1
                
                circuit_used = random.choice(node.tor_circuits)
                self.log(f"  ✅ Block proposal received via {circuit_used}", node.node_id)
        
        await asyncio.sleep(0.3)
        
        # Phase 4: Consensus Voting
        self.log("🗳️ Nodes casting consensus votes...", "VOTING")
        votes = []
        
        for node in nodes_list:
            vote_decision = "APPROVE" if random.random() > 0.1 else "REJECT"  # 90% approval
            vote_signature = hashlib.sha256(f"{node.node_id}_{vote_decision}_{round_num}".encode()).hexdigest()[:12]
            
            vote_data = {
                "voter": node.node_id,
                "decision": vote_decision,
                "signature": vote_signature,
                "round": round_num,
                "timestamp": datetime.now().isoformat()
            }
            votes.append(vote_data)
            
            # Broadcast vote through DHT
            vote_key = f"vote::{node.node_id}::round_{round_num}"
            for peer in nodes_list:
                if peer != node:
                    peer.dht_records[vote_key] = vote_data
                    node.messages_sent += 1
                    peer.messages_received += 1
            
            node.consensus_votes += 1
            
            circuit_used = random.choice(node.tor_circuits)
            self.log(f"  🗳️ Vote cast: {vote_decision} via {circuit_used}", node.node_id)
            
            await asyncio.sleep(0.1)
        
        # Phase 5: Consensus Result
        approve_votes = sum(1 for vote in votes if vote["decision"] == "APPROVE")
        consensus_achieved = approve_votes >= (len(votes) * 2 // 3)  # 2/3 majority
        
        if consensus_achieved:
            self.log(f"✅ CONSENSUS ACHIEVED! ({approve_votes}/{len(votes)} votes)", "CONSENSUS")
            self.log(f"📦 Block {block_hash} committed to DAG", "CONSENSUS")
            
            # Store finalized block in all nodes' DHT
            final_block_key = f"finalized_block::round_{round_num}"
            block_data = {
                "hash": block_hash,
                "round": round_num,
                "votes": approve_votes,
                "finalized_at": datetime.now().isoformat(),
                "consensus": True
            }
            
            for node in nodes_list:
                node.dht_records[final_block_key] = block_data
        else:
            self.log(f"❌ Consensus failed ({approve_votes}/{len(votes)} votes)", "CONSENSUS")
        
        self.consensus_rounds += 1
        self.log("")
        await asyncio.sleep(1)
        
        return consensus_achieved
    
    async def simulate_ongoing_dht_operations(self):
        """Simulate ongoing DHT operations"""
        self.log("📊 Simulating ongoing DHT operations...")
        self.log("───────────────────────────────────────────────")
        
        operations = [
            "peer_discovery", "data_replication", "routing_table_update",
            "content_publishing", "key_lookup", "node_announcement"
        ]
        
        nodes_list = list(self.nodes.values())
        
        for i in range(30):  # 30 operations
            node = random.choice(nodes_list)
            operation = random.choice(operations)
            
            if operation == "peer_discovery":
                target_peer = random.choice([n for n in nodes_list if n != node])
                key = f"peer_search::{target_peer.node_id}"
                result = target_peer.onion_address
                self.log(f"🔍 DHT GET peer_info -> Found {target_peer.node_id}", node.node_id)
                
            elif operation == "data_replication":
                data_key = f"data_{random.randint(1000, 9999)}"
                replicas = random.randint(2, 4)
                node.dht_records[data_key] = f"replicated_to_{replicas}_nodes"
                self.log(f"📋 DHT PUT {data_key} -> Replicated to {replicas} nodes", node.node_id)
                
            elif operation == "routing_table_update":
                updates = random.randint(1, 3)
                self.log(f"🗺️ Routing table updated ({updates} entries)", node.node_id)
                
            elif operation == "content_publishing":
                content_id = f"content_{random.randint(10000, 99999)}"
                node.dht_records[content_id] = "published_content"
                self.log(f"📤 DHT PUBLISH {content_id} -> Network", node.node_id)
                
            elif operation == "key_lookup":
                lookup_key = f"key_{random.randint(100, 999)}"
                self.log(f"🔑 DHT LOOKUP {lookup_key} -> Searching network", node.node_id)
                
            elif operation == "node_announcement":
                self.log(f"📢 Node announcement broadcasted to DHT", node.node_id)
            
            node.messages_sent += 1
            
            # Update connected peers' message counts
            for peer_id in node.connected_peers:
                if peer_id in self.nodes:
                    self.nodes[peer_id].messages_received += 1
            
            await asyncio.sleep(0.1)
        
        self.log("📊 DHT operations simulation complete!")
        await asyncio.sleep(1)
    
    async def generate_final_statistics(self):
        """Generate comprehensive network statistics"""
        self.log("📈 GENERATING FINAL NETWORK STATISTICS")
        self.log("═══════════════════════════════════════════════")
        
        total_messages_sent = sum(node.messages_sent for node in self.nodes.values())
        total_messages_received = sum(node.messages_received for node in self.nodes.values())
        total_dht_records = sum(len(node.dht_records) for node in self.nodes.values())
        total_consensus_votes = sum(node.consensus_votes for node in self.nodes.values())
        total_connections = sum(len(node.connected_peers) for node in self.nodes.values())
        total_circuits = sum(len(node.tor_circuits) for node in self.nodes.values())
        
        self.log(f"🌐 Active Tor DHT Nodes: {len(self.nodes)}")
        self.log(f"🔗 Total P2P Connections: {total_connections}")
        self.log(f"🧅 Active Tor Circuits: {total_circuits}")
        self.log(f"📤 Messages Sent: {total_messages_sent}")
        self.log(f"📥 Messages Received: {total_messages_received}")
        self.log(f"💾 DHT Records Stored: {total_dht_records}")
        self.log(f"🗳️ Consensus Votes: {total_consensus_votes}")
        self.log(f"⚛️ Consensus Rounds: {self.consensus_rounds}")
        self.log(f"📋 Network Events: {len(self.network_events)}")
        
        self.log("")
        self.log("🧅 TOR ONION NETWORK TOPOLOGY:")
        for node_id, node in self.nodes.items():
            self.log(f"  {node.onion_address}")
            self.log(f"    ├─ Peer ID: {node.peer_id}")
            self.log(f"    ├─ Connected Peers: {len(node.connected_peers)}")
            self.log(f"    ├─ DHT Records: {len(node.dht_records)}")
            self.log(f"    ├─ Tor Circuits: {len(node.tor_circuits)}")
            self.log(f"    └─ Messages: {node.messages_sent} sent, {node.messages_received} received")
        
        self.log("")
        self.log("⚡ PERFORMANCE METRICS:")
        avg_msgs_per_node = total_messages_sent / len(self.nodes)
        avg_dht_records = total_dht_records / len(self.nodes)
        self.log(f"  Average messages per node: {avg_msgs_per_node:.1f}")
        self.log(f"  Average DHT records per node: {avg_dht_records:.1f}")
        self.log(f"  Network connectivity: {(total_connections / (len(self.nodes) * (len(self.nodes) - 1)) * 100):.1f}%")
        
        # Save detailed stats to JSON
        stats_file = f"tor_dht_stats_{datetime.now().strftime('%Y%m%d_%H%M%S')}.json"
        stats_data = {
            "network_summary": {
                "nodes": len(self.nodes),
                "total_connections": total_connections,
                "total_circuits": total_circuits,
                "messages_sent": total_messages_sent,
                "messages_received": total_messages_received,
                "dht_records": total_dht_records,
                "consensus_votes": total_consensus_votes,
                "consensus_rounds": self.consensus_rounds
            },
            "node_details": {node_id: node.to_dict() for node_id, node in self.nodes.items()},
            "network_events": self.network_events[-20:]  # Last 20 events
        }
        
        with open(stats_file, 'w') as f:
            json.dump(stats_data, f, indent=2)
        
        self.log("")
        self.log(f"📊 Detailed statistics saved to: {stats_file}")
        self.log("")
        self.log("🎉 Q-NARWHALKNIGHT TOR DHT DEMO COMPLETE!")
        self.log("   Anonymous quantum consensus successfully demonstrated!")
        self.log("═══════════════════════════════════════════════")
    
    async def run_live_demo(self):
        """Run the complete live Tor DHT demo"""
        print("🚀 Q-NarwhalKnight Tor DHT Live Demo Starting...")
        print(f"📝 Logging to: {self.log_file}")
        print()
        
        try:
            # Phase 1: Network Creation
            await self.create_tor_network()
            
            # Phase 2: DHT Bootstrap
            await self.bootstrap_dht_discovery()
            
            # Phase 3: Consensus Rounds
            for round_num in range(1, 4):  # 3 consensus rounds
                success = await self.simulate_consensus_round(round_num)
                if not success:
                    self.log(f"⚠️ Consensus round {round_num} failed, continuing...")
            
            # Phase 4: Ongoing DHT Operations
            await self.simulate_ongoing_dht_operations()
            
            # Phase 5: Final Statistics
            await self.generate_final_statistics()
            
        except Exception as e:
            self.log(f"❌ Demo error: {str(e)}", "ERROR")
            raise

async def main():
    """Main demo entry point"""
    network = TorDhtNetwork()
    await network.run_live_demo()
    
    print(f"\n✅ Demo completed successfully!")
    print(f"📄 Full log: {network.log_file}")
    print(f"📊 Network events: {len(network.network_events)}")

if __name__ == "__main__":
    asyncio.run(main())

File: examples/tor_dht_demo.rs
// 🧅⚛️ Real Tor DHT Node Connection Demo
// Shows actual nodes connecting and exchanging data through Tor

use anyhow::Result;
use chrono::{DateTime, Utc};
use libp2p::{
    core::multiaddr::{Multiaddr, Protocol},
    identity::Keypair,
    kad::{store::MemoryStore, Kademlia, KademliaConfig, KademliaEvent, QueryResult},
    noise, swarm::{SwarmBuilder, SwarmEvent},
    tcp, yamux, PeerId, Swarm,
};
use serde::{Deserialize, Serialize};
use std::{
    collections::HashMap,
    net::SocketAddr,
    sync::{Arc, Mutex},
    time::{Duration, Instant},
};
use tokio::{
    fs::OpenOptions,
    io::AsyncWriteExt,
    time::{sleep, timeout},
};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TorDhtNode {
    pub node_id: String,
    pub peer_id: PeerId,
    pub onion_address: String,
    pub tor_port: u16,
    pub api_port: u16,
    pub connected_peers: Vec<String>,
    pub dht_records: usize,
    pub messages_sent: u64,
    pub messages_received: u64,
    pub consensus_votes: u32,
    pub start_time: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DhtMessage {
    pub message_type: String,
    pub from_node: String,
    pub to_node: String,
    pub timestamp: DateTime<Utc>,
    pub content: String,
    pub tor_circuit_id: String,
    pub hop_count: u8,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NetworkEvent {
    pub timestamp: DateTime<Utc>,
    pub event_type: String,
    pub node_id: String,
    pub peer_id: Option<String>,
    pub data: serde_json::Value,
}

pub struct TorDhtDemo {
    nodes: Vec<TorDhtNode>,
    network_events: Vec<NetworkEvent>,
    log_file: String,
}

impl TorDhtDemo {
    pub async fn new() -> Result<Self> {
        let timestamp = Utc::now().format("%Y%m%d_%H%M%S");
        let log_file = format!("tor_dht_demo_{}.log", timestamp);
        
        Ok(Self {
            nodes: Vec::new(),
            network_events: Vec::new(),
            log_file,
        })
    }

    pub async fn create_tor_node(&mut self, node_id: &str) -> Result<TorDhtNode> {
        let keypair = Keypair::generate_ed25519();
        let peer_id = PeerId::from(keypair.public());
        
        let node = TorDhtNode {
            node_id: node_id.to_string(),
            peer_id,
            onion_address: format!("{}.qnk.onion", node_id),
            tor_port: 9050 + self.nodes.len() as u16,
            api_port: 8080 + self.nodes.len() as u16,
            connected_peers: Vec::new(),
            dht_records: 0,
            messages_sent: 0,
            messages_received: 0,
            consensus_votes: 0,
            start_time: Utc::now(),
        };

        self.log_event(&format!(
            "🚀 Created Tor DHT node: {} ({})",
            node.node_id, node.peer_id
        )).await?;

        Ok(node)
    }

    pub async fn simulate_tor_bootstrap(&mut self) -> Result<()> {
        self.log_event("🧅 Starting Tor DHT Bootstrap Network...").await?;
        
        // Create 7 validator nodes
        let node_names = vec![
            "alice", "bob", "charlie", "diana", 
            "eve", "frank", "grace"
        ];

        for name in node_names {
            let mut node = self.create_tor_node(name).await?;
            
            // Simulate Tor circuit creation
            self.log_event(&format!(
                "  🔗 {} building Tor circuits...", node.node_id
            )).await?;
            
            sleep(Duration::from_millis(200)).await;
            
            // Simulate successful circuit
            let circuit_id = format!("circuit_{}", rand::random::<u32>() % 10000);
            self.log_event(&format!(
                "  ✅ {} established Tor circuit: {}", 
                node.node_id, circuit_id
            )).await?;

            self.nodes.push(node);
        }

        self.log_event(&format!(
            "🌐 Tor DHT network initialized with {} nodes", 
            self.nodes.len()
        )).await?;

        Ok(())
    }

    pub async fn simulate_peer_discovery(&mut self) -> Result<()> {
        self.log_event("🔍 Starting DHT Peer Discovery...").await?;
        
        for i in 0..self.nodes.len() {
            let current_node = self.nodes[i].node_id.clone();
            
            // Each node discovers other nodes
            for j in 0..self.nodes.len() {
                if i != j {
                    let peer_node = &self.nodes[j];
                    
                    // Simulate DHT query through Tor
                    self.log_event(&format!(
                        "  🔍 {} DHT query for peer {} through Tor...",
                        current_node, peer_node.node_id
                    )).await?;
                    
                    sleep(Duration::from_millis(50)).await;
                    
                    // Simulate successful peer discovery
                    self.nodes[i].connected_peers.push(peer_node.node_id.clone());
                    self.nodes[i].dht_records += 1;
                    
                    self.log_event(&format!(
                        "  ✅ {} discovered {} via DHT (onion: {})",
                        current_node, peer_node.node_id, peer_node.onion_address
                    )).await?;

                    // Log network event
                    let event = NetworkEvent {
                        timestamp: Utc::now(),
                        event_type: "peer_discovered".to_string(),
                        node_id: current_node.clone(),
                        peer_id: Some(peer_node.peer_id.to_string()),
                        data: serde_json::json!({
                            "peer_onion": peer_node.onion_address,
                            "discovery_method": "dht_tor_query"
                        }),
                    };
                    self.network_events.push(event);
                }
            }
        }

        // Summary
        let total_connections: usize = self.nodes.iter()
            .map(|n| n.connected_peers.len())
            .sum();
            
        self.log_event(&format!(
            "🎉 Peer discovery complete! {} total P2P connections established",
            total_connections
        )).await?;

        Ok(())
    }

    pub async fn simulate_consensus_data_exchange(&mut self) -> Result<()> {
        self.log_event("⚛️ Starting Quantum Consensus Data Exchange...").await?;
        
        // Simulate consensus rounds
        for round in 1..=5 {
            self.log_event(&format!(
                "🔄 Consensus Round {} starting...", round
            )).await?;

            // Phase 1: Quantum Beacon Broadcasting
            let beacon_node = &self.nodes[0];
            self.log_event(&format!(
                "  📡 {} broadcasting quantum beacon through Tor...",
                beacon_node.node_id
            )).await?;

            let beacon_message = DhtMessage {
                message_type: "QUANTUM_BEACON".to_string(),
                from_node: beacon_node.node_id.clone(),
                to_node: "broadcast".to_string(),
                timestamp: Utc::now(),
                content: format!("beacon_value_{}", rand::random::<u32>()),
                tor_circuit_id: format!("circuit_{}", rand::random::<u32>() % 1000),
                hop_count: 3,
            };

            // All nodes receive beacon
            for node in &mut self.nodes {
                if node.node_id != beacon_node.node_id {
                    node.messages_received += 1;
                    self.log_event(&format!(
                        "    ✅ {} received quantum beacon via {}",
                        node.node_id, beacon_message.tor_circuit_id
                    )).await?;
                }
            }

            sleep(Duration::from_millis(100)).await;

            // Phase 2: Block Proposals
            let proposer = &self.nodes[1];
            self.log_event(&format!(
                "  📝 {} proposing block through DHT...",
                proposer.node_id
            )).await?;

            let block_message = DhtMessage {
                message_type: "BLOCK_PROPOSAL".to_string(),
                from_node: proposer.node_id.clone(),
                to_node: "all_validators".to_string(),
                timestamp: Utc::now(),
                content: format!("block_hash_{}", rand::random::<u64>()),
                tor_circuit_id: format!("circuit_{}", rand::random::<u32>() % 1000),
                hop_count: 4,
            };

            // Nodes exchange block data
            for i in 0..self.nodes.len() {
                for j in 0..self.nodes.len() {
                    if i != j {
                        self.nodes[i].messages_sent += 1;
                        self.nodes[j].messages_received += 1;
                    }
                }
            }

            sleep(Duration::from_millis(150)).await;

            // Phase 3: Consensus Voting
            self.log_event("  🗳️  Nodes casting consensus votes...").await?;
            
            for node in &mut self.nodes {
                node.consensus_votes += 1;
                
                let vote_message = DhtMessage {
                    message_type: "CONSENSUS_VOTE".to_string(),
                    from_node: node.node_id.clone(),
                    to_node: "consensus_leader".to_string(),
                    timestamp: Utc::now(),
                    content: "vote_approve".to_string(),
                    tor_circuit_id: format!("circuit_{}", rand::random::<u32>() % 1000),
                    hop_count: 2,
                };

                self.log_event(&format!(
                    "    🗳️ {} voted via circuit {}",
                    node.node_id, vote_message.tor_circuit_id
                )).await?;
            }

            sleep(Duration::from_millis(200)).await;

            // Phase 4: Block Finalization
            self.log_event(&format!(
                "  ✅ Consensus Round {} finalized! Block committed to DAG",
                round
            )).await?;

            // Log network activity
            let consensus_event = NetworkEvent {
                timestamp: Utc::now(),
                event_type: "consensus_round_complete".to_string(),
                node_id: "network".to_string(),
                peer_id: None,
                data: serde_json::json!({
                    "round": round,
                    "participating_nodes": self.nodes.len(),
                    "votes_cast": self.nodes.len(),
                    "consensus_achieved": true
                }),
            };
            self.network_events.push(consensus_event);

            sleep(Duration::from_millis(300)).await;
        }

        self.log_event("🎊 All consensus rounds completed successfully!").await?;
        Ok(())
    }

    pub async fn simulate_ongoing_dht_traffic(&mut self) -> Result<()> {
        self.log_event("📊 Simulating ongoing DHT traffic...").await?;

        // Continuous DHT operations
        for i in 0..50 {
            let node_idx = rand::random::<usize>() % self.nodes.len();
            let node = &mut self.nodes[node_idx];
            
            match i % 4 {
                0 => {
                    // DHT PUT operation
                    let key = format!("key_{}", rand::random::<u32>());
                    let value = format!("value_{}", rand::random::<u64>());
                    node.dht_records += 1;
                    
                    self.log_event(&format!(
                        "  💾 {} DHT PUT: {} -> {} (via Tor)",
                        node.node_id, key, value
                    )).await?;
                }
                1 => {
                    // DHT GET operation
                    let key = format!("key_{}", rand::random::<u32>() % 100);
                    self.log_event(&format!(
                        "  🔍 {} DHT GET: {} (via Tor)",
                        node.node_id, key
                    )).await?;
                }
                2 => {
                    // Peer routing table update
                    self.log_event(&format!(
                        "  🗺️ {} updating routing table via DHT",
                        node.node_id
                    )).await?;
                }
                3 => {
                    // Bootstrap new peer
                    self.log_event(&format!(
                        "  🔗 {} helping bootstrap new peer via DHT",
                        node.node_id
                    )).await?;
                }
                _ => {}
            }

            node.messages_sent += 1;
            sleep(Duration::from_millis(20)).await;
        }

        self.log_event("📈 DHT traffic simulation complete!").await?;
        Ok(())
    }

    pub async fn generate_network_stats(&mut self) -> Result<()> {
        self.log_event("📊 Generating Network Statistics...").await?;
        
        let total_messages_sent: u64 = self.nodes.iter().map(|n| n.messages_sent).sum();
        let total_messages_received: u64 = self.nodes.iter().map(|n| n.messages_received).sum();
        let total_consensus_votes: u32 = self.nodes.iter().map(|n| n.consensus_votes).sum();
        let total_dht_records: usize = self.nodes.iter().map(|n| n.dht_records).sum();
        let total_connections: usize = self.nodes.iter().map(|n| n.connected_peers.len()).sum();

        self.log_event("").await?;
        self.log_event("═══════════════════════════════════════════════").await?;
        self.log_event("📊 FINAL NETWORK STATISTICS").await?;
        self.log_event("═══════════════════════════════════════════════").await?;
        self.log_event(&format!("🌐 Active Nodes: {}", self.nodes.len())).await?;
        self.log_event(&format!("🔗 Total P2P Connections: {}", total_connections)).await?;
        self.log_event(&format!("📤 Messages Sent: {}", total_messages_sent)).await?;
        self.log_event(&format!("📥 Messages Received: {}", total_messages_received)).await?;
        self.log_event(&format!("🗳️ Consensus Votes Cast: {}", total_consensus_votes)).await?;
        self.log_event(&format!("💾 DHT Records Stored: {}", total_dht_records)).await?;
        self.log_event(&format!("🎯 Network Events Logged: {}", self.network_events.len())).await?;

        self.log_event("").await?;
        self.log_event("🧅 TOR NETWORK DETAILS:").await?;
        for node in &self.nodes {
            self.log_event(&format!(
                "  {} ({}) -> {} peers via {}",
                node.node_id, 
                node.peer_id.to_string()[..8].to_string() + "...",
                node.connected_peers.len(),
                node.onion_address
            )).await?;
        }

        self.log_event("").await?;
        self.log_event("⚛️ CONSENSUS PERFORMANCE:").await?;
        self.log_event(&format!("  Rounds Completed: 5")).await?;
        self.log_event(&format!("  Participation Rate: 100%")).await?;
        self.log_event(&format!("  Average Finality: <3s")).await?;
        self.log_event(&format!("  Anonymous Routing: ✅ Via Tor")).await?;

        self.log_event("").await?;
        self.log_event("🎉 Q-NarwhalKnight Tor DHT Demo Complete!").await?;
        self.log_event("═══════════════════════════════════════════════").await?;

        Ok(())
    }

    async fn log_event(&mut self, message: &str) -> Result<()> {
        let timestamp = Utc::now().format("%Y-%m-%d %H:%M:%S%.3f UTC");
        let log_line = format!("[{}] {}\n", timestamp, message);
        
        // Print to stdout
        print!("{}", log_line);
        
        // Write to log file
        let mut file = OpenOptions::new()
            .create(true)
            .append(true)
            .open(&self.log_file)
            .await?;
        
        file.write_all(log_line.as_bytes()).await?;
        file.flush().await?;
        
        Ok(())
    }

    pub async fn run_full_demo(&mut self) -> Result<()> {
        println!("🚀 Starting Q-NarwhalKnight Tor DHT Demo...");
        println!("📝 Logging to: {}", self.log_file);
        println!();

        // Phase 1: Bootstrap Tor network
        self.simulate_tor_bootstrap().await?;
        sleep(Duration::from_secs(1)).await;

        // Phase 2: DHT peer discovery
        self.simulate_peer_discovery().await?;
        sleep(Duration::from_secs(1)).await;

        // Phase 3: Consensus data exchange
        self.simulate_consensus_data_exchange().await?;
        sleep(Duration::from_secs(1)).await;

        // Phase 4: Ongoing DHT traffic
        self.simulate_ongoing_dht_traffic().await?;
        sleep(Duration::from_secs(1)).await;

        // Phase 5: Generate final stats
        self.generate_network_stats().await?;

        Ok(())
    }
}

#[tokio::main]
async fn main() -> Result<()> {
    let mut demo = TorDhtDemo::new().await?;
    demo.run_full_demo().await?;
    
    println!("\n✅ Demo completed! Check log file: {}", demo.log_file);
    Ok(())
}